net_connection/src/net_connection_thread.rs

```diff
@@ -0,0 +1,183 @@
+use super::NetResult;
+
+use super::{
+    net_connection::{NetConnection, NetHandler, NetWorkerFactory},
+    protocol::Protocol,
+};
+
+use std::{thread, time};
+
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    mpsc, Arc,
+};
+
+/// a NetConnection instance that is managed on another thread
+#[derive(Debug)]
+pub struct NetConnectionThread {
+    keep_running: Arc<AtomicBool>,
+    send_channel: mpsc::Sender<Protocol>,
+    thread: thread::JoinHandle<()>,
+}
+
+impl NetConnection for NetConnectionThread {
+    /// send a message to the worker within this NetConnectionThread instance
+    fn send(&mut self, data: Protocol) -> NetResult<()> {
+        self.send_channel.send(data)?;
+        Ok(())
+    }
+}
+
+impl NetConnectionThread {
+    /// stop (join) the worker thread
+    pub fn stop(self) -> NetResult<()> {
+        self.keep_running.store(false, Ordering::Relaxed);
+        match self.thread.join() {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                bail!("NetConnectionThread failed to join on stop() call");
+            }
+        }
+    }
+
+    /// create a new NetConnectionThread instance with given handler / worker
+    pub fn new(handler: NetHandler, mut worker_factory: NetWorkerFactory) -> NetResult<Self> {
+        let keep_running = Arc::new(AtomicBool::new(true));
+        let keep_running2 = keep_running.clone();
+
+        let (sender, receiver) = mpsc::channel();
+        Ok(NetConnectionThread {
+            keep_running,
+            send_channel: sender,
+            thread: thread::spawn(move || {
+                let mut us = 100_u64;
+                let mut worker = match worker_factory(handler) {
+                    Ok(w) => w,
+                    Err(e) => panic!("{:?}", e),
+                };
+
+                while keep_running2.load(Ordering::Relaxed) {
+                    let mut did_something = false;
+
+                    match receiver.try_recv() {
+                        Ok(data) => {
+                            did_something = true;
+                            match worker.receive(data) {
+                                Ok(_) => (),
+                                Err(e) => panic!("{:?}", e),
+                            };
+                        }
+                        Err(_) => (),
+                    };
```

Comment:
```rust
receiver.try_recv().and_then(|data| {
    did_something = true;
    worker.receive(data)
        .unwrap_or_else(|e| panic!("{:?}", e));
}).unwrap_or(());
```

Reply:
@sphinxc0re - happy to make these changes, just curious about the rational. Is it mainly stylistic? From my perspective, I always feel like direct code is preferable to closures even if the compiler will probably optimize them to the same thing...

Reply:
updated

Reply:
Because it makes the control flow more clear. If you are using a function that takes a closure, you can't `return` from the outer function and thus it's easier to grasp the control flow directly.